lib/cadet/accounts/luminus.ex

```
@@ -0,0 +1,211 @@
+  """
+
+  @api_key :cadet |> Application.fetch_env!(:updater) |> Keyword.get(:luminus_api_key)
+  @client_id :cadet |> Application.fetch_env!(:updater) |> Keyword.get(:luminus_client_id)
+  @client_secret :cadet |> Application.fetch_env!(:updater) |> Keyword.get(:luminus_client_secret)
+  @redirect_url :cadet |> Application.fetch_env!(:updater) |> Keyword.get(:luminus_redirect_url)
+  @api_token_url "https://luminus.nus.edu.sg/v2/auth/connect/token"
+  @api_url "https://luminus.azure-api.net/"
+
+  # Construct a valid URL with the module attributes, and given params
+  # The authentication token parameter must be provided explicitly rather than
+  # provided implicitly by this function as some API calls use ...&Token={token},
+  # while others use ...&AuthToken={token}
+  defp api_url(method, queries) do
+    @api_url
+    |> URI.merge(method)
+    |> Map.put(:query, URI.encode_query(queries))
+    |> URI.to_string()
+  end
+
+  @doc """
+  Exchange OAuth Authorization code for a valid access token.
+  Token will expire in 30 minutes from the time it was issued.
+
+  returns...
+
+    - {:ok, :token} - valid token, token is a string
+    - {:error, :bad_request} - invalid token
+    - {:error, :internal_server_error} - the luminus_client_secret is invalid
+
+  ## Parameters
+
+    - code: String, the LumiNUS authorization code
+
+  ## Examples
+
+      iex> Cadet.Accounts.LumiNUS.fetch_luminus_token("C0dE...")
+      {:ok, "12742174091894830298409823098098"}
+  """
+
+  def fetch_luminus_token(code) do
+    queries =
+      %{
+        client_id: @client_id,
+        client_secret: @client_secret,
+        grant_type: "authorization_code",
+        code: code,
+        redirect_uri: @redirect_url
+      }
+      |> URI.encode_query()
+
+    headers = [{"Content-type", "application/x-www-form-urlencoded"}]
+
+    case HTTPoison.post(@api_token_url, queries, headers) do
+      {:ok, %{body: body, status_code: 200}} when body != ~s("") ->
```

Why do you need the guard clause `when body != ~s("")`?

Just let it crash!